src/accountsdb/cache.zig

```diff
@@ -0,0 +1,209 @@
+        account: Account,
+        ref_count: ReferenceCounter,
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) error{OutOfMemory}!CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = .{},
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            if (self.ref_count.release()) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            }
+        }
+
+        // satisfies the type LruCacheCustom expects
+        pub fn releaseOrDestroyDoublePtr(self: **CachedAccount, allocator: std.mem.Allocator) void {
+            releaseOrDestroy(self.*, allocator);
+        }
+
+        pub fn copyRef(self: *CachedAccount) *CachedAccount {
+            if (self.ref_count.acquire()) {
+                return self;
+            } else {
+                @panic("Attempted to acquire reference to destroyed CachedAccount");
+            }
+        }
+
+        fn getMeta(self: *CachedAccount) CachedAccountMeta {
+            return .{
+                .ref_count = self.ref_count,
+            };
+        }
+    };
+
+    /// A copy of a CachedAccout's fields, without the account
+    const CachedAccountMeta = struct {
+        ref_count: ReferenceCounter,
+    };
+
+    pub const LRU = LruCacheCustom(
+        .locking,
+        Pubkey,
+        *CachedAccount,
+        std.mem.Allocator,
+        CachedAccount.releaseOrDestroyDoublePtr,
+    );
+
+    const Self = @This();
+
+    pub fn init(
+        allocator: std.mem.Allocator,
+        max_items: usize,
+    ) error{OutOfMemory}!AccountsCache {
+        return .{
+            .lru = try LRU.initWithContext(allocator, max_items, allocator),
+            .allocator = allocator,
+        };
+    }
+
+    /// User is expected to release the returned CachedAccount.
+    pub fn get(self: *Self, pubkey: Pubkey) ?*CachedAccount {
+        const account: *CachedAccount = self.lru.get(pubkey) orelse return null;
+        return account.copyRef();
+    }
+
+    /// Returns a copy of the CachedAccount's meta. Likely shouldn't be used outside of tests.
+    fn getMeta(self: *Self, pubkey: Pubkey) ?CachedAccountMeta {
+        const account: *CachedAccount = self.lru.peek(pubkey) orelse return null;
+        return account.getMeta();
+    }
+
+    /// Does not influence underlying LRU's cache ordering
+    pub fn contains(self: *const Self, pubkey: Pubkey) bool {
+        return self.lru.contains(pubkey);
+    }
+
+    /// Replaces the previous entry, if one existed
+    pub fn put(self: *Self, pubkey: Pubkey, account: Account) error{OutOfMemory}!void {
+        const new_entry = try self.allocator.create(CachedAccount);
```

Comment:
This allocation seems superfluous to me. The LRU already allocates new keys when they are added. Can we rewrite something somewhere to not need this second allocation on every `put` instead of on the first one?

Reply:
The reasoning is that the LRU doesn't necessarily "own" the account, we don't necessarily want the entry to be deallocated when it's evicted; references to it could still be held. With that being said we could remove one allocation per put by allocating a CachedAccount and its .account field in the same block, but that feels like a bit of a hack.

Reply:
I think the core issue here is that the reference counter must be allocated in some memory that is not owned by the lru. It's fine to copy the CachedAccount struct into the Lru but it needs to respect the same reference counter that's being used by other copies of the CachedAccount.

So you need an allocation outside the Lru regardless, it's just a question of whether to allocate the CachedAccount as a whole, or to only allocate its reference counter.

Reply:
I went ahead and changed this to an unreachable, I think conveys intent a bit better